timemachine/cpp/src/kernels/k_harmonic_bond.cuh

```diff
@@ -13,52 +13,49 @@ void __global__ k_harmonic_bond(
     unsigned long long *__restrict__ du_dp,
     __int128 *__restrict__ u) {
 
-    const auto b_idx = blockDim.x * blockIdx.x + threadIdx.x;
+    int b_idx = blockDim.x * blockIdx.x + threadIdx.x;
 
-    if (b_idx >= B) {
-        return;
-    }
-
-    int src_idx = bond_idxs[b_idx * 2 + 0];
-    int dst_idx = bond_idxs[b_idx * 2 + 1];
+    while (b_idx < B) {
+        int src_idx = bond_idxs[b_idx * 2 + 0];
+        int dst_idx = bond_idxs[b_idx * 2 + 1];
 
-    RealType dx[3];
-    RealType d2ij = 0;
-    for (int d = 0; d < 3; d++) {
-        RealType delta = coords[src_idx * 3 + d] - coords[dst_idx * 3 + d];
-        dx[d] = delta;
-        d2ij += delta * delta;
-    }
+        RealType dx[3];
+        RealType d2ij = 0;
+#pragma unroll
+        for (int d = 0; d < 3; d++) {
+            RealType delta = coords[src_idx * 3 + d] - coords[dst_idx * 3 + d];
+            dx[d] = delta;
+            d2ij += delta * delta;
+        }
 
-    int kb_idx = b_idx * 2 + 0;
-    int b0_idx = b_idx * 2 + 1;
+        int kb_idx = b_idx * 2 + 0;
+        int b0_idx = b_idx * 2 + 1;
 
-    RealType kb = params[kb_idx];
-    RealType b0 = params[b0_idx];
+        RealType kb = params[kb_idx];
+        RealType b0 = params[b0_idx];
 
-    RealType dij = sqrt(d2ij);
-    RealType db = dij - b0;
+        RealType dij = sqrt(d2ij);
+        RealType db = dij - b0;
 
-    if (du_dx) {
-        for (int d = 0; d < 3; d++) {
-            RealType grad_delta;
-            if (b0 != 0) {
-                grad_delta = kb * db * dx[d] / dij;
-            } else {
-                grad_delta = kb * dx[d];
+        if (du_dx) {
+            const RealType inv_dij = 1 / dij;
```

Comment:
curiosity: what's the motivation for pulling this out as a subexpression when it's only used once?

Reply:
Its used 3 times within the for loop below

Reply:
Sorry, I missed that it was used outside of the diff